In [27]:
import pandas as pd

# Работа с Pandas

Загрузите ваш DataFrame в переменную df и выведете на экран первые пять строк 

In [1]:
path = r'G:\Мой диск\Colab Notebooks\1T\ДЗ_1T\2\googleplaystore.csv'# Считываем наш датафрейм из csv файла
df = pd.read_csv(path, encoding='utf-8')
df.fillna(0, axis=1, inplace=True)

NameError: name 'pd' is not defined

Отобразите количество строк и столбцов в DataFrame

In [28]:
a,b = df.shape
print(f'число строк в DF = {a},\nчисло столбцов в DF = {b}')

число строк в DF = 10841,
число столбцов в DF = 13


# PRO часть (Не обязательна к выполнению, но желательна :))


In [29]:
df.head(0)

Empty DataFrame
Columns: [App, Category, Rating, Reviews, Size, Installs, Type, Price, Content Rating, Genres, Last Updated, Current Ver, Android Ver]
Index: []

Измените названия всех столбцов приведя их к нижнему регистру

In [30]:
# Комбинируем метод .replace(что меняем, на что меняем) и метод .lower() приводим к нижнему регистру
# В качестве аргумента в .rename передаем лямбда-функцию (в которой входной параметр х - это названия каждого столбца)
df.rename(columns=lambda x: x.replace(" ", "_").lower(), inplace=True) 
df.head(0)

Empty DataFrame
Columns: [app, category, rating, reviews, size, installs, type, price, content_rating, genres, last_updated, current_ver, android_ver]
Index: []

Выведите основные статистики по категориальным столбцам. Напишите выводы о них 

In [31]:
df.describe(include='object') # тип "object" обычно используется для хранения категориальных данных

app category  rating reviews                size    installs  \
count    10841    10841   10841   10841               10841       10841   
unique    9660       34      41    6002                 462          22   
top     ROBLOX   FAMILY       0       0  Varies with device  1,000,000+   
freq         9     1972    1474     596                1695        1579   

         type    price content_rating genres    last_updated  \
count   10841  10841.0          10841  10841           10841   
unique      4     92.0              7    120            1378   
top      Free      0.0       Everyone  Tools  August 3, 2018   
freq    10039  10041.0           8714    842             326   

               current_ver android_ver  
count                10841       10841  
unique                2833          34  
top     Varies with device  4.1 and up  
freq                  1459        2451

In [32]:
categorical_columns = df.select_dtypes(include=['object']).columns 
# используем метод `.select_dtypes()` с аргументом `include=['object']`, чтобы выбрать только те столбцы, которые имеют тип "object"

for category in categorical_columns:            # принято решение обойти в цикле каждую категориальную колонку
    print("Статистика для столбца -", category)
    print(df[category].value_counts())          # выводим количество уникальных значений в столбце
    print(df[category].describe())              # выводим описательные статистики для столбца
    print(f'{"-"*75}')

Статистика для столбца - app
app
ROBLOX                                                9
CBS Sports App - Scores, News, Stats & Watch Live     8
ESPN                                                  7
Duolingo: Learn Languages Free                        7
Candy Crush Saga                                      7
                                                     ..
Meet U - Get Friends for Snapchat, Kik & Instagram    1
U-Report                                              1
U of I Community Credit Union                         1
Waiting For U Launcher Theme                          1
iHoroscope - 2018 Daily Horoscope & Astrology         1
Name: count, Length: 9660, dtype: int64
count      10841
unique      9660
top       ROBLOX
freq           9
Name: app, dtype: object
---------------------------------------------------------------------------
Статистика для столбца - category
category
FAMILY                 1972
GAME                   1144
TOOLS                   843
MEDICAL        

    ВЫВОДЫ:

В столбце category присутсвует выброс (1.9 - 1 - строка 56 вывода)
В основной своей массе отзывы по приложениям отсутствуют, но есть приложение с 398307 отзывов.
Удивительно, но есть 58 приложений с кол-ом скачиваний 1,000,000,000+ 
Интересно, что судя по столбцу last_updated 3 Августа 2018г було обновлено сразу - 326 приложений 
Самое топовое приложение - ROBLOX
Самая топовая категория - FAMILY
Люди больше всего скачивают бесплатные прииложения.

Выведите на экран часть датафрейма, который соответствует условию: `приложение из категории искусства и дизайна` И `с рейтингом больше 4.5`

In [34]:
df.loc[(df['category'] == 'ART_AND_DESIGN') & (df['rating'] > 4.5)]
#  df['category'] == 'ART_AND_DESIGN' - работаем по Булевой маске с условием нахождения в столбце category словосочетания ART_AND_DESIGN
#  df['rating'] > 4.5 - тоже создает Булевую маску где True при  условии рейтинга > 4.5
#  & - объединяем две Булевые серии операндом И
#  передаем все это методу .loc для отбора и вывода строк содержащих True end True

app        category  \
2     U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN   
9                         Kids Paint Free - Drawing Fun  ART_AND_DESIGN   
13                                Mandala Coloring Book  ART_AND_DESIGN   
16         Photo Designer - Write your name with shapes  ART_AND_DESIGN   
19                                         ibis Paint X  ART_AND_DESIGN   
22              Superheroes Wallpapers | 4K Backgrounds  ART_AND_DESIGN   
24                          HD Mickey Minnie Wallpapers  ART_AND_DESIGN   
25                           Harley Quinn wallpapers HD  ART_AND_DESIGN   
26                        Colorfit - Drawing & Coloring  ART_AND_DESIGN   
34                                      I Creative Idea  ART_AND_DESIGN   
36       UNICORN - Color By Number & Pixel Art Coloring  ART_AND_DESIGN   
38                       PIP Camera - PIP Collage Maker  ART_AND_DESIGN   
45    Canva: Poster, banner, card maker & graphic de...  ART_AND_DESIGN   
46    Install images with music to make video withou...  ART_AND_DESIGN   
3982                                  Cardi B Wallpaper  ART_AND_DESIGN   
4749  X Launcher: With OS11 Style Theme & Control Ce...  ART_AND_DESIGN   
4755  X Launcher Pro: PhoneX Theme, OS11 Control Center  ART_AND_DESIGN   
4759  X Launcher Pro - IOS Style Theme & Control Center  ART_AND_DESIGN   
4764     X Launcher Prime: With OS Style Theme & No Ads  ART_AND_DESIGN   
5254                            AJ Styles HD Wallpapers  ART_AND_DESIGN   
6930               Fantasy theme dark bw black building  ART_AND_DESIGN   
8888            Spring flowers theme couleurs d t space  ART_AND_DESIGN   

     rating reviews  size     installs  type price content_rating  \
2       4.7   87510  8.7M   5,000,000+  Free   0.0       Everyone   
9       4.7     121  3.1M      10,000+  Free   0.0       Everyone   
13      4.6    4326   21M     100,000+  Free   0.0       Everyone   
16      4.7    3632  5.5M     500,000+  Free   0.0       Everyone   
19      4.6  224399   31M  10,000,000+  Free   0.0       Everyone   
22      4.7    7699  4.2M     500,000+  Free   0.0   Everyone 10+   
24      4.7     118   23M      50,000+  Free   0.0       Everyone   
25      4.8     192  6.0M      10,000+  Free   0.0       Everyone   
26      4.7   20260   25M     500,000+  Free   0.0       Everyone   
34      4.7     353  4.2M      10,000+  Free   0.0           Teen   
36      4.7    8145   24M     500,000+  Free   0.0       Everyone   
38      4.7     158   11M      10,000+  Free   0.0       Everyone   
45      4.7  174531   24M  10,000,000+  Free   0.0       Everyone   
46      4.6    1070   26M     100,000+  Free   0.0       Everyone   
3982    4.8     253  3.7M      50,000+  Free   0.0       Everyone   
4749    4.7    5754  4.4M     100,000+  Free   0.0       Everyone   
4755    4.7     801  3.5M       5,000+  Paid  1.99       Everyone   
4759    4.8    1216  8.6M      10,000+  Paid  1.99       Everyone   
4764    4.7     149  3.5M       1,000+  Paid  1.99       Everyone   
5254    4.8      48   25M       5,000+  Free   0.0       Everyone   
6930    4.8      41  1.9M       5,000+  Free   0.0       Everyone   
8888    5.0       1  2.9M         100+  Free   0.0       Everyone   

                       genres       last_updated current_ver   android_ver  
2                Art & Design     August 1, 2018       1.2.4  4.0.3 and up  
9     Art & Design;Creativity       July 3, 2018         2.8  4.0.3 and up  
13               Art & Design      June 26, 2018       1.0.4    4.4 and up  
16               Art & Design      July 31, 2018         3.1    4.1 and up  
19               Art & Design      July 30, 2018       5.5.4    4.1 and up  
22               Art & Design      July 12, 2018     2.2.6.2  4.0.3 and up  
24               Art & Design       July 7, 2018       1.1.3    4.1 and up  
25               Art & Design     April 25, 2018         1.5    3.0 and up  
26    Art & Design;Creativity   October 11, 2

Посчитатайте какое количество игр содержится в рейтинге, используя параметр разбиения (bins) равным 100. Выведите на экран топ-10 игр по количеству в рейтинге. Напишите краткий вывод.
То есть сначала отфильтровываете лишь игры из всего датафрейма, после чего применяете value_counts к столбцу с рейтингом используя параметр, который указан выше. И выводите первые 10 строк отсортированных по убыванию.

In [ ]:
df['rating'] = df['rating'].astype(float)  # Преобразование столбца в числовой формат

In [72]:
df.loc[(df['category'] == 'GAME')].value_counts()

app                                               category  rating  reviews   size                installs      type  price  content_rating  genres     last_updated       current_ver         android_ver       
Clash of Clans                                    GAME      4.6     44891723  98M                 100,000,000+  Free  0.0    Everyone 10+    Strategy   July 15, 2018      10.322.16           4.1 and up            2
Sniper 3D Gun Shooter: Free Shooting Games - FPS  GAME      4.6     7674252   Varies with device  100,000,000+  Free  0.0    Mature 17+      Action     August 2, 2018     Varies with device  Varies with device    2
Bowmasters                                        GAME      4.7     1535973   Varies with device  50,000,000+   Free  0.0    Teen            Action     July 23, 2018      2.12.5              4.1 and up            2
Flow Free                                         GAME      4.3     1295625   11M                 100,000,000+  Free  0.0    Everyone        Puzz

В рейтинге:  1121 игры

In [74]:
df['rating'].value_counts(bins=100).head(10)

rating
(4.37, 4.56]     2147
(4.18, 4.37]     2028
(-0.02, 0.19]    1474
(4.56, 4.75]     1322
(3.99, 4.18]     1276
(3.61, 3.8]       542
(3.8, 3.99]       386
(3.42, 3.61]      337
(4.75, 4.94]      321
(4.94, 5.13]      274
Name: count, dtype: int64

In [77]:
df.loc[(df['category'] == 'GAME')].nlargest(10, 'rating')

app category  rating reviews  \
5795                              Axe Champs! Wars     GAME     5.0       8   
5822   Flippy Axe : Flip The Knife & Axe Simulator     GAME     5.0       7   
5917                                      Ra Ga Ba     GAME     5.0       2   
6700                              Brick Breaker BR     GAME     5.0       7   
7402                          Trovami se ci riesci     GAME     5.0      11   
7466                                        211:CK     GAME     5.0       8   
7698                                     CP Trivia     GAME     5.0       5   
9056                   Santa's Monster Shootout DX     GAME     5.0       4   
9265                                      EC Mover     GAME     5.0       5   
10697                                      Mu.F.O.     GAME     5.0       2   

       size installs  type price content_rating  genres        last_updated  \
5795    25M      50+  Free   0.0       Everyone  Arcade       June 26, 2018   
5822    15M     100+  Free   0.0       Everyone  Arcade  September 27, 2017   
5917    20M       1+  Paid  1.49       Everyone  Arcade    February 8, 2017   
6700    19M       5+  Free   0.0       Everyone  Arcade       July 23, 2018   
7402   6.1M      10+  Free   0.0       Everyone  Arcade      March 11, 2017   
7466    38M      10+  Paid  0.99           Teen  Arcade      April 11, 2018   
7698    12M     100+  Free   0.0       Everyone  Trivia      August 6, 2018   
9056    33M      50+  Paid  1.99           Teen  Action     August 15, 2013   
9265   4.6M      10+  Free   0.0       Everyone  Racing      August 1, 2018   
10697   16M       1+  Paid  0.99       Everyone  Arcade       March 3, 2017   

      current_ver   android_ver  
5795          1.1    4.1 and up  
5822          1.1    4.2 and up  
5917        1.0.4    2.3 and up  
6700          1.0    4.1 and up  
7402          0.1    2.3 and up  
7466          1.3    4.1 and up  
7698         0.99    5.0 and up  
9056         1.05    2.2 and up  
9265         1.11  4.0.3 and up  
10697         1.0    2.3 and up

    Выводы: 
- почти все приложения, аркадного типа (где надо просто тыкать, все мелькает и собирают монетки, убиватели t)
- Почти половина из рейтинга - это платные приложения.
- только  2 из 10 игр в категории Teen, остальные в Everyone

Создайте функцию , которая будет применена к столбцу рейтинг методом `apply()`.
Функция должна принимать строку с рейтингом и возвращать:
 - High rating, если рейтинг больше или равен 4.5
 - Middle rating, если рейтинг между 3.8 и 4.5
 - Low rating, если рейтинг меньше 3.8 
Результат работы функции запишите в новый столбец с названием `categorical_rating`

In [78]:
def level_rating(x):
    if x >= 4.5:
        return "High rating"
    elif x < 3.8:
        return "Low rating"
    else:
        return "Middle rating"

In [80]:
df['categorical_rating'] = df['rating'].apply(level_rating)
df.head(5)

app        category  rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  reviews  size     installs  type price content_rating  \
0     159   19M      10,000+  Free   0.0       Everyone   
1     967   14M     500,000+  Free   0.0       Everyone   
2   87510  8.7M   5,000,000+  Free   0.0       Everyone   
3  215644   25M  50,000,000+  Free   0.0           Teen   
4     967  2.8M     100,000+  Free   0.0       Everyone   

                      genres      last_updated         current_ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    android_ver categorical_rating  
0  4.0.3 and up      Middle rating  
1  4.0.3 and up      Middle rating  
2  4.0.3 and up        High rating  
3    4.2 and up        High rating  
4    4.4 and up      Middle rating

Выведите кол-во уникальных значений в столбцах отсортированных по убыванию на экран:

In [81]:
unique_counts = df.nunique().sort_values(ascending=False)
print(unique_counts)

app                   9660
reviews               6002
current_ver           2833
last_updated          1378
size                   462
genres                 120
price                   92
rating                  41
category                34
android_ver             34
installs                22
content_rating           7
type                     4
categorical_rating       3
dtype: int64


__Super-Hard задания__

Решите задачу, где мы создавали свою собственную функцию, но вместо функции `def` используйте `лямбда-функцию` внутри `apply()`

In [82]:
df['categorical_rating'] = df['rating'].apply(lambda x: "High rating" if x >= 4.5 else ("Low rating" if x < 3.8 else "Middle rating"))
df.head(5)

app        category  rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  reviews  size     installs  type price content_rating  \
0     159   19M      10,000+  Free   0.0       Everyone   
1     967   14M     500,000+  Free   0.0       Everyone   
2   87510  8.7M   5,000,000+  Free   0.0       Everyone   
3  215644   25M  50,000,000+  Free   0.0           Teen   
4     967  2.8M     100,000+  Free   0.0       Everyone   

                      genres      last_updated         current_ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    android_ver categorical_rating  
0  4.0.3 and up      Middle rating  
1  4.0.3 and up      Middle rating  
2  4.0.3 and up        High rating  
3    4.2 and up        High rating  
4    4.4 and up      Middle rating

Изучите функцию `transform` в документации Pandas. Примените её к датафрейму, который будет заранее сгруппирован по столбцу `category`. С помощью фукнции `transform` найдите средний `rating` по каждой категории и запишите в столбец с названием `mean_cat_rating`. Кол-во строк в получившемся новом столбец должно быть таким же, как и в исходном датафрейме

    ссылка на transform

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.transform.html

In [92]:
df['mean_cat_rating'] = df.groupby('category')['rating'].transform('mean')
# выполняем группировку по столбцу category
# применяем агрегирующую функцию .transform('mean') к столбцу rating в каждой группе
#### Функция `transform` используется для применения агрегирующей функции к группам данных 
#### сохранения результата в новом столбце с тем же количеством строк, что и исходный DataFrame.

df.head(4)

app        category  rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   

  reviews  size     installs  type price content_rating  \
0     159   19M      10,000+  Free   0.0       Everyone   
1     967   14M     500,000+  Free   0.0       Everyone   
2   87510  8.7M   5,000,000+  Free   0.0       Everyone   
3  215644   25M  50,000,000+  Free   0.0           Teen   

                      genres      last_updated         current_ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   

    android_ver categorical_rating  mean_cat_rating  
0  4.0.3 and up      Middle rating         4.156923  
1  4.0.3 and up      Middle rating         4.156923  
2  4.0.3 and up        High rating         4.156923  
3    4.2 and up        High rating         4.156923

Создайте функцию , которая будет применена к столбцу `last updated` и возвращать из него число месяца, если число месяца вернуть из строки невозможно, то верните вместо этого строку с значением `miss_date`.\
Запишите полученный столбец в датафрейм , название столбцу дайте `day_of_update`.

Если вы столкнётесь с некоторой ошибкой , которую вам выдаст Pandas, то изучите конструкцию Python `try` \ `except`.\
Вот псевдокод , который поможет лучше разобраться с этой конструкцией:


Привет!

Давайте разберем код по шагам:

Шаг 1: Вы определяете функцию `extract_day_of_update`, которая принимает аргумент `date`.
```python
def extract_day_of_update(date):
```

Шаг 2: Внутри функции есть блок try-except. В блоке try выполняется следующее:
- Функция `pd.to_datetime(date)` преобразует значение `date` в объект типа `datetime`.
- Затем мы используем `.day`, чтобы получить только день месяца из этого объекта типа `datetime`.
- Если успешно выполнены все эти шаги, то возвращаем день месяца.

Если что-то происходит не так, тогда переходим к блоку except и возвращаем строку `'miss_date'`.
```python
    try:
        return pd.to_datetime(date).day
    except:
        return 'miss_date'
```

Шаг 3: Вы применяете функцию `extract_day_of_update` к столбцу `'last_updated'` вашего датафрейма с помощью метода `.apply()` и сохраняете результаты в новом столбце `'day_of_update'`.
```python
df['day_of_update'] = df['last_updated'].apply(extract_day_of_update)
```

Шаг 4: Вы выводите первые 5 строк измененного датафрейма с помощью функции `display` (либо `print`, если вы используете обычный питоновский интерпретатор).
```python
display(df.head(5))
```

Надеюсь, это помогло! Если у вас возникнут еще вопросы, пожалуйста, не стесняйтесь спрашивать.


In [97]:
def extract_day_of_update(date):         # Функция `pd.to_datetime(date)` преобразует значение `date` в объект типа `datetime`.
    try:                                 # Затем мы используем `.day`, чтобы получить только день месяца из этого объекта типа `datetime`.
        return pd.to_datetime(date).day  # Если успешно выполнены все эти шаги, то возвращаем день месяца.
    except:
        return 'miss_date'               # Если что-то происходит не так return 'miss_date'

df['day_of_update'] = df['last_updated'].apply(extract_day_of_update)
display(df.head(5))

app        category  rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  reviews  size     installs  type price content_rating  \
0     159   19M      10,000+  Free   0.0       Everyone   
1     967   14M     500,000+  Free   0.0       Everyone   
2   87510  8.7M   5,000,000+  Free   0.0       Everyone   
3  215644   25M  50,000,000+  Free   0.0           Teen   
4     967  2.8M     100,000+  Free   0.0       Everyone   

                      genres      last_updated         current_ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    android_ver categorical_rating  mean_cat_rating day_of_update  
0  4.0.3 and up      Middle rating         4.156923             7  
1  4.0.3 and up      Middle rating         4.156923            15  
2  4.0.3 and up        High rating         4.156923             1  
3    4.2 and up        High rating         4.156923             8  
4    4.4 and up      Middle rating         4.156923            20

-- Интересно что данный код выполняется 2.5 секунды.. 